testing/trino-tests/src/test/java/io/trino/execution/TestQueryTracker.java

```diff
@@ -0,0 +1,92 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.execution;
+
+import com.google.common.collect.ImmutableList;
+import io.trino.Session;
+import io.trino.connector.MockConnectorFactory;
+import io.trino.spi.Plugin;
+import io.trino.spi.connector.ColumnMetadata;
+import io.trino.spi.connector.ConnectorFactory;
+import io.trino.testing.DistributedQueryRunner;
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+import static io.trino.SystemSessionProperties.QUERY_MAX_PLANNING_TIME;
+import static io.trino.spi.type.VarcharType.VARCHAR;
+import static io.trino.testing.TestingSession.testSessionBuilder;
+import static java.lang.Integer.MAX_VALUE;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+// Tests relies on timeouts so additional query runners may disturb
+@Test(singleThreaded = true)
+public class TestQueryTracker
+{
+    private DistributedQueryRunner queryRunner;
+
+    @BeforeMethod
+    public void setup()
+            throws Exception
+    {
+        Session defaultSession = testSessionBuilder()
+                .setCatalog("mock")
+                .setSchema("default")
+                .setSystemProperty(QUERY_MAX_PLANNING_TIME, "5s")
+                .build();
+
+        queryRunner = DistributedQueryRunner
+                .builder(defaultSession)
+                .build();
+        queryRunner.installPlugin(new Plugin()
+        {
+            @Override
+            public Iterable<ConnectorFactory> getConnectorFactories()
+            {
+                return ImmutableList.of(MockConnectorFactory.builder()
+                        .withGetColumns(ignored -> ImmutableList.of(new ColumnMetadata("name", VARCHAR)))
+                        // Apply Join happens inside optimizer so this should model most blocking tasks in planning phase
+                        .withApplyJoin((ignored1, ignored2, ignored3, ignored4, ignored5, ignored6, ignored7) -> freeze())
+                        .build());
+            }
+        });
+        queryRunner.createCatalog("mock", "mock");
+    }
+
+    @AfterMethod(alwaysRun = true)
+    public void tearDown()
+    {
+        queryRunner.close();
+        queryRunner = null;
+    }
+
+    @Test(timeOut = 10_000)
+    public void testInterruptFetchingStatistics()
+    {
+        assertThatThrownBy(() -> queryRunner.execute("SELECT count(*) FROM t1 n1 JOIN t2 n2 ON n1.name = n2.name"))
+                .hasMessageContaining("Query exceeded the maximum planning time limit of 5.00s");
+    }
+
+    private static <T> T freeze()
+    {
+        try {
+            Thread.sleep(MAX_VALUE);
+        }
+        catch (InterruptedException e) {
+            throw new RuntimeException(e);
```

Comment:
set a future on interrupted exception and validate that it was set in `testInterruptFetchingStatistics` after planning interrupted

Reply:
What about future there?